lalrpop/src/build/mod.rs

```diff
@@ -205,42 +207,29 @@ fn needs_rebuild(lalrpop_file: &Path, rs_file: &Path) -> io::Result<bool> {
 
 fn lalrpop_files<P: AsRef<Path>>(root_dir: P) -> io::Result<Vec<PathBuf>> {
     let mut result = vec![];
-    for entry in fs::read_dir(root_dir)? {
+
+    let walkdir = WalkDir::new(root_dir)
+        .follow_links(true)
+        // Use deterministic ordering:
+        .sort_by_file_name()
+        // Prevent loops:
+        .same_file_system(true);
```

Comment:
"Currently, this option is only supported on Unix and Windows. If this option is used on an unsupported platform, then directory traversal will immediately return an error and will not yield any entries." -> Is this something to be worried about?

Reply:
Perhaps it's best to just accept non-termination in case of loops resulting from bind mounts then, since `same_file_system` would also break some legitimate (i.e. directed) mount setups? I've removed the `same_file_system` call for now, let me know if I should put it in again.

Reply:
It's something I'm fine with, if `walkdir` ever addresses this in a better way then we can add it back in. @dburgener can weigh in if he has other ideas.

Reply:
> Is this something to be worried about?

I did some digging to try to determine what "unix" meant in this case.  The reason it's exclusive to "unix" and "windows" is these calls to WalkDir's private "device_num" function https://docs.rs/crate/walkdir/latest/source/src/util.rs.  That is conditionally compiled on `target_family` `unix` or `windows`, and errors otherwise.

The `target_family` config is extremely poorly documented.  Most of the links point to https://doc.rust-lang.org/reference/conditional-compilation.html, which gives "unix", "windows", and "wasm" as examples, but doesn't tell you how OSs map to families, or whether that list of examples is exclusive.

Eventually I found this stackoverflow post: https://stackoverflow.com/questions/66845716/where-are-rusts-built-in-target-specifications-defined which gives a way to query the architecture to `target_family` mapping on your system.  Running that command locally confirms that `unix`, `windows` and `wasm` are the only three options for built in target specs on my system.

So my understanding is that that "unix" in this context means all unix-like operating systems (eg Linux, BSD (including Mac), android, etc).  So "unix" and "windows" is pretty close to covering everything.  The only other mainstream option is "wasm".  Since this is code to detect lalrpop files at build time, that would break someone who's building directly in a wasm environment (which might actually be the situation for the rust playground?).  Of course, users aren't limited to the built-in options.  In theory a user could set target_family to an arbitary string, which would probably happen in specific embedded case.  Again though, this is called from build.rs, so if you're cross-compiling to a custom target, you wouldn't hit this code.

In other words, I suspect that use cases without "unix" or "windows" here, while technically possible are probably rare (the rust playground might be the big exception though).

One option could be to do a `#[cfg(any(unix, windows))]` ourselves, just to be certain, with the behavior of setting the same_file_system option if we're on one of those families, and otherwise not.

All that said, maybe if you've created a looped directory structure via bind mounts inside your source tree, you only have yourself to blame when things go wrong, so it may not be worth it.

Reply:
No strong opinions, but I think your final point is accurate. Lalrpop should do the reasonable thing which works for the most amount of people and if your file system is cursed enough then it's up to you to figure out a work around.